src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,228 @@
+  SplineProblem<> splineProblem(
+    Eigen::Vector2d(_currTime, _nextTime), _numCoefficients, numDofs);
+
+  assert(_currPosition.size() == numDofs);
+  assert(_nextPosition.size() == numDofs);
+  splineProblem.addConstantConstraint(0, 0, _currPosition);
+  splineProblem.addConstantConstraint(1, 0, _nextPosition);
+
+  if (_numCoefficients >= 4)
+  {
+    assert(_currVelocity.size() == numDofs);
+    assert(_nextVelocity.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 1, _currVelocity);
+    splineProblem.addConstantConstraint(1, 1, _nextVelocity);
+  }
+
+  if (_numCoefficients >= 6)
+  {
+    assert(_currAcceleration.size() == numDofs);
+    assert(_nextAcceleration.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 2, _currAcceleration);
+    splineProblem.addConstantConstraint(1, 2, _nextAcceleration);
+  }
+
+  const auto splineSegment = splineProblem.fit();
+  return splineSegment.getCoefficients()[0];
+}
+
+//=============================================================================
+void extractJointTrajectoryPoint(
+  const trajectory_msgs::JointTrajectory& _trajectory,
+  size_t _index, size_t _numDofs,
+  Eigen::VectorXd* _positions, bool _positionsRequired,
+  Eigen::VectorXd* _velocities, bool _velocitiesRequired,
+  Eigen::VectorXd* _accelerations, bool _accelerationsRequired)
+{
+  const auto& waypoint = _trajectory.points[_index];
+
+  try
+  {
+    checkVector("positions", waypoint.positions, _numDofs,
+      _positionsRequired, _positions);
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, _velocities);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& _space,
+  const trajectory_msgs::JointTrajectory& _jointTrajectory)
+{
+  if (!_space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledJoints = _space->getNumSubspaces();
+  if (_jointTrajectory.joint_names.size() != numControlledJoints)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledJoints << ", got "
+        << _jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  // Check that all joints are single DOF.
+  for (size_t i = 0; i < _space->getNumSubspaces(); ++i)
+  {
+    auto n = _space->getJointSpace(i)->getJoint()->getNumDofs();
+    if (n != 1)
+    {
+      std::stringstream message;
+      message << "[Conversion] Expected 1 dof. Joint "
```

Comment:
Nit: Only this warning message has the heading `[Conversion]`. It would good to be consistent by either removing the heading from here or adding the heading to all other messages.